cmd/node-agent/main.go

```diff
@@ -0,0 +1,278 @@
+	client = newAPIClient()
+	nodename, err := getNodeName()
+	if err != nil {
+		glog.Fatal(err)
+	}
+
+	glog.Infof("starting node agent, watching node: %s", nodename)
+
+	sysdConn, err = dbus.New()
+	if err != nil {
+		glog.Fatal(err)
+	}
+
+	nodeopts := api.ListOptions{
+		FieldSelector: fields.ParseSelectorOrDie(fmt.Sprintf("metadata.name=%s", nodename)),
+	}
+	_, nodeController := framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(lo api.ListOptions) (runtime.Object, error) {
+				return client.Core().Nodes().List(nodeopts)
+			},
+			WatchFunc: func(lo api.ListOptions) (watch.Interface, error) {
+				return client.Core().Nodes().Watch(nodeopts)
+			},
+		},
+		&v1.Node{},
+		30*time.Second,
+		framework.ResourceEventHandlerFuncs{
+			UpdateFunc: nodeUpdateCallback,
+		},
+	)
+	var configMapController *framework.Controller
+	configMapStore, configMapController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(lo api.ListOptions) (runtime.Object, error) {
+				return client.Core().ConfigMaps(api.NamespaceSystem).List(lo)
+			},
+			WatchFunc: func(lo api.ListOptions) (watch.Interface, error) {
+				return client.Core().ConfigMaps(api.NamespaceSystem).Watch(lo)
+			},
+		},
+		&v1.ConfigMap{},
+		30*time.Second,
+		framework.ResourceEventHandlerFuncs{},
+	)
+
+	go configMapController.Run(wait.NeverStop)
+	nodeController.Run(wait.NeverStop)
+}
+
+func newAPIClient() clientset.Interface {
+	config, err := restclient.InClusterConfig()
+	if err != nil {
+		glog.Fatal(err)
+	}
+	return clientset.NewForConfigOrDie(config)
+}
+
+func nodeUpdateCallback(oldObj, newObj interface{}) {
+	glog.Infof("begin node update callback")
+
+	newNode, ok := newObj.(*v1.Node)
+	if !ok {
+		glog.Errorf("received unexpected type: %T (expected *v1.Node)", newObj)
+		return
+	}
+	if newNode.Annotations == nil {
+		newNode.Annotations = make(map[string]string)
+	}
+	rawconf, ok := newNode.Annotations[desiredConfigAnnotation]
+	if !ok {
+		glog.Info("no configuration annotation found, ignoring")
+		return
+	}
+	var desiredConf map[string]string
+	err := json.Unmarshal([]byte(rawconf), &desiredConf)
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+
+	onDiskConfig, err := parseKubeletEnvFile(kubeletEnvPath)
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+	cm, err := getConfigMap(desiredConf[kubeletConfigKey])
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+	if configHasChanged(onDiskConfig, desiredConf, cm) {
+		if err := validateConfig(desiredConf); err != nil {
+			glog.Error(err)
+			return
+		}
+		if err := updateKubeletEnvFile(desiredConf, onDiskConfig, cm); err != nil {
+			glog.Error(err)
+			return
+		}
+		if err := restartKubeletService(); err != nil {
+			glog.Error(err)
+			return
+		}
+	}
+	// Always update annotation at end of sync loop.
+	if err := updateNodeAnnotation(newNode, desiredConf); err != nil {
+		glog.Error(err)
+	}
+}
+
+func configHasChanged(onDiskConfig, desiredConfig map[string]string, cm *v1.ConfigMap) bool {
```

Comment:
nit: add a godoc description of what this function checks. Mentally parsing something like `onDiskConfig[kubeletFlagsKey] != cm.Data[configMapFlagsKey]` isn't immediately clear.